compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4381,16 +4381,23 @@ private BType checkOptionalRecordFieldAccessExpr(BLangFieldBasedAccess fieldAcce
 
         LinkedHashSet<BType> fieldTypeMembers = new LinkedHashSet<>();
 
+        boolean nonMatchedRecordExists = false;
+
         for (BType memType : memberTypes) {
             BType individualFieldType = checkOptionalRecordFieldAccessExpr(fieldAccessExpr, memType, fieldName);
 
             if (individualFieldType == symTable.semanticError) {
+                nonMatchedRecordExists = true;
                 continue;
             }
 
             fieldTypeMembers.add(individualFieldType);
         }
 
+        if (nonMatchedRecordExists) {
+            fieldTypeMembers.add(symTable.nilType);
```

Comment:
Shall we rename https://github.com/ballerina-platform/ballerina-lang/blob/8b26828e30589512d938031ecc115c03a1a57800/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L4255 and use that instead?

Reply:
Done